dali/operators/nvcvop/nvcvop.h

```diff
@@ -0,0 +1,222 @@
+                                     TensorLayout layout = "") {
+    auto dtype = nvcv::DataType(DTYPE);
+    int num_samples = ws.GetInputBatchSize(0);
+    int arg_sample_vol = volume(arg_shape);
+    arg.Acquire(spec_, ws, num_samples, arg_shape);
+    T *staging_buffer =
+        scratchpad.AllocatePinned<T>(num_samples * arg_sample_vol, dtype.alignment());
+    for (int s = 0; s < num_samples; ++s) {
+      std::copy(arg[s].data, arg[s].data + arg_sample_vol,
+                staging_buffer + s * arg_sample_vol);
+    }
+    T *device_buffer = scratchpad.AllocateGPU<T>(num_samples * arg_sample_vol, dtype.alignment());
+    MemCopy(device_buffer, staging_buffer, num_samples * arg_sample_vol * sizeof(T), ws.stream());
+
+    auto shape_data = CalcTensorShape(arg_shape, num_samples, dtype);
+    nvcv::TensorDataStridedCuda::Buffer inBuf;
+    inBuf.basePtr = reinterpret_cast<NVCVByte*>(device_buffer);
+    inBuf.strides[shape_data.size() - 1] = dtype.strideBytes();
+    for (int d = shape_data.size() - 2; d >= 0; --d) {
+      inBuf.strides[d] = shape_data[d + 1] * inBuf.strides[d + 1];
+    }
+    TensorLayout out_layout = layout.empty() ? layout : "N" + layout;
+    nvcv::TensorShape shape(shape_data.data(), shape_data.size(),
+                            nvcv::TensorLayout(out_layout.c_str()));
+    nvcv::TensorDataStridedCuda inData(shape, dtype, inBuf);
+    return nvcv::TensorWrapData(inData);
+  }
+
+  std::vector<int64_t> CalcTensorShape(const TensorShape<> &arg_shape, int num_samples,
+                                       const nvcv::DataType &dtype) {
+    int ndim = arg_shape.sample_dim();
+    int64_t inner_dim = arg_shape[ndim - 1];
+    DALI_ENFORCE(dtype.numChannels() == 1 || inner_dim == dtype.numChannels(),
+                 make_string("Invalid argument shape. Inner dimension should match "
+                 "the number of channels in the data type: ", dtype.numChannels()));
+    std::vector<int64_t> shape_data(ndim + 1);
+    shape_data[0] = num_samples;
+    for (int d = 0; d < ndim; ++d) {
+      shape_data[d + 1] = arg_shape[d];
+    }
+    // If number of channels in the data type is greater than 1,
+    // we remove the inner dimension from the DALI shape
+    if (dtype.numChannels() > 1) {
+      shape_data.pop_back();
+    }
+    return shape_data;
+  }
+
+  /**
+   * @brief Get input image batch.
+   * This method allocates the ImageBatchVarShape and wraps the input data as nvcv images.
+   *
+   * The output should be used only within the RunImpl method.
+   */
+  const nvcv::ImageBatchVarShape &GetInputBatch(Workspace &ws, size_t input_idx) {
+    if (inputs_.size() < input_idx + 1) {
+      inputs_.resize(input_idx + 1, nvcv::ImageBatchVarShape{1});
+    }
+    const auto &tl_input = ws.Input<GPUBackend>(input_idx);
+    auto &input = inputs_[input_idx];
+    if (input.capacity() < tl_input.num_samples()) {
+      input = nvcv::ImageBatchVarShape(std::max(input.capacity() * 2, tl_input.num_samples()));
```

Comment:
Just to leave a mark on what we've discussed offline - do we need or consider making the ImageBatchVarShape internal allocations handled by DALI allocators/scratchpad? With some aggresive prefetching scenarios it could(?) trade more memory usage for less synchronization, but I don't know if we have actual need for that.

Reply:
I think we should wrap our allocators into an NVCV Allocator. Perhaps this can be done in a follow-up, but it definitely _should_ be done.